### ISSUE TITLE
What is the expected behavior after all the tasks are done?

### ISSUE THREAD

**Issue by nathantippy** [vertexclique/nuclei#37]
Here is a modified version of the file write example.  After it is done running I sleep for 34 seconds but if we monitor the task manager one core remains pegged.  Is this desired behavior or is there a way to release the pool after it has been engaged?

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#[nuclei::main]
async fn main() -> io::Result<()> {
    {
        // Approximately ~75,9 MB
        let dark_matter = vec![DARK_MATTER_TEXT; 100_000].join("\n");

        let path = PathBuf::from("testfile.txt");

        let fo = OpenOptions::new()
            .create(true)
            .write(true)
            .open(&path)
            .unwrap();
        let mut file = Handle::<File>::new(fo).unwrap();
        file.write_all(dark_matter.as_bytes()).await.unwrap();
        let _ = file.close().await;

        let fo = OpenOptions::new()
            .read(true)
            .open(&path)
            .unwrap();

        let mut file = Handle::<File>::new(fo).unwrap();

        let mut buf = vec![];
        assert!(file.seek(SeekFrom::Start(0)).await.is_ok());
        assert_eq!(file.read_to_end(&mut buf).await.unwrap(), dark_matter.len());
        assert_eq!(&buf[0..dark_matter.len()], dark_matter.as_bytes());
        let _ = file.close().await;

        println!("Length of file is {}", buf.len());
    }
    //We have no more work to do yet we have 1 core pegged at 100%.
    Delay::new(Duration::from_secs(34)).await;

    Ok(())
}

**Comment by vertexclique** [comment]
Please take a look at https://docs.rs/nuclei/latest/nuclei/config/struct.IoUringConfiguration.html#impl-IoUringConfiguration
Also it would be good to take a look at this example:
https://github.com/vertexclique/nuclei/blob/master/examples/proactor-config-fwrite.rs#L27

**Comment by nathantippy** [comment]
Thanks, that is just what I was looking for.  After the drive() call the CPU drops back down however if I call drive() again later it pegs the CPU and stays there.  Here is my adjusted proactor-config-fwrite showing the issue.  I also noted in the comment that when I tried io_poll I saw an assert failure::   

fn main() -> io::Result<()> {
    let nuclei_config = NucleiConfig {
        iouring: IoUringConfiguration::interrupt_driven(1 << 11), //pops up to 100% after second drive(

        //iouring: IoUringConfiguration::io_poll(1 << 11), // assertion failed: mid <= self.len()
        //thread 'main' panicked at /home/nate/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/io/write_all.rs:33:58:
        //assertion failed: mid <= self.len()

    };
    let _ = Proactor::with_config(nuclei_config);

    let mut count = 3;
    while count > 0 {

            drive(async {
                // Approximately ~75,9 MB
                let dark_matter = vec![DARK_MATTER_TEXT; 100_000].join("\n");
                println!("writing file");
                let mut path = PathBuf::from("testfile.txt");

                let fo = OpenOptions::new()
                    .create(true)
                    .read(true)
                    .write(true)
                    .open(&path)
                    .unwrap();
                let mut file = Handle::<File>::new(fo).unwrap();
                file.write_all(dark_matter.as_bytes()).await.unwrap();

                let mut buf = vec![];
                assert!(file.seek(SeekFrom::Start(0)).await.is_ok());
                assert_eq!(file.read_to_end(&mut buf).await.unwrap(), dark_matter.len());
                assert_eq!(&buf[0..dark_matter.len()], dark_matter.as_bytes());

                file.close().await;
                println!("Length of file is {}", buf.len());

            });
        sleep(Duration::from_secs(3));
        count -= 1;
    }
    sleep(Duration::from_secs(32));

    Ok(())
}

**Comment by vertexclique** [comment]
Ok, drive is needed to be called once of a lifetime of a program. If you continuously call it will exhaust both kernel and userland.
Should have added that to docs.
If you want to run async code then please use spawn, spawn_blocking, spawn_local etc.

IO_POLL works on files opened with O_DIRECT. It is kind of hard to detect incoming handles conforming to this.

So if you register wrong combinations for uring it will fail at registering ring buffers.

**Comment by vertexclique** [comment]
I am using server grade for testing please reduce ring size to smth like 256 or 128. That is why it is failing at the registry.

**Comment by nathantippy** [comment]
I get it, but how should I integrate this with something like bastion?  should I set up one child actor just to do my IO so drive() belongs there OR should I put all of the bastion child tree building inside drive() so any actor can make use of this? Also, if someone takes my crate and they call drive() how would they know that I already used it once? Seems like it should be behind some kind of lock.

**Comment by vertexclique** [comment]
One child driving the io with drive is good.

nuclei::drive(futures::pending())
Smth like this^^^

The rest of the supervisor tree can stay the same.

It is possible to make IO drive run only once, but mind that if someone wants to register multiple times a backend, it won't be possible. I will figure out a way to add an alternative for multiple backends.

**Comment by nathantippy** [comment]
Excellent, thanks for the help and feedback. I look forward to your new features.

**Comment by nathantippy** [comment]
Surprisingly this also pegs one core even before we have any work and even If I have async sleep inside of drive().


    let nuclei_config = NucleiConfig {
        iouring: IoUringConfiguration::interrupt_driven(1 << 6),
        //iouring: IoUringConfiguration::kernel_poll_only(1 << 6),
    };
    let _ = nuclei::Proactor::with_config(nuclei_config);    

    nuclei::spawn_blocking(|| {
        nuclei::drive(pending::<()>());

        //nuclei::drive(async {
        //    loop { Delay::new(Duration::from_secs(1)).await; }
        //});
    }).detach();

**Comment by nathantippy** [comment]
turns out putting a normal sleep(Duration::from_secs(3));  in that loop did the trick.

**Comment by nathantippy** [comment]
Also I needed to move drive() out to main  because if drive() is in a bastion child and gets restarted we are back in the same boat calling drive() again each time the child is restarted by the super.

**Comment by vertexclique** [comment]
> Also I needed to move drive() out to main because if drive() is in a bastion child and gets restarted

True, that's the way.

> turns out putting a normal sleep(Duration::from_secs(3)); in that loop did the trick.

~This can block the main thread, be careful. But if that solved for you, that's good. <3~

Ok, that code looks correct. I didn't see that.
